Create Ca-certificate configmap  used by token projected volume

WanLinghao opened the pull request:
To inject ca.crt into container when projected volume was specified, configmap should be created in each namespace.This patch add a controller called "ca-configmap" to complete above job as well as some bootstrap rbac policies.

fix: https://github.com/kubernetes/kubernetes/issues/68811
```release-note
NONE
```

liggitt commented:
I expected these pieces to be added by the admission plugin, and the backing configmap for the CA to be created by the controller manager, which already has that info

/assign @mikedanese

WanLinghao commented:
I got something pending:

- Do we need to add labels to the configmaps created in each namespace like current serviceaccount-secret do?

- We could just manage the configmaps via existed serviceaccount controller. Or build a independent one to do configmaps manage job. I think the independent one is better idea since it could keep the code clear and reduce the clean work we need to do when we completely abandon  current serviceaccount model.
WDYT @mikedanese @liggitt

WanLinghao commented:
/retest

WanLinghao commented:
Sorry, not "add labels", but add something like SecretType which could distinguish if the configmap was created by controller automatically.
> I got something pending:
> 
>     * Do we need to add labels to the configmaps created in each namespace like current serviceaccount-secret do?

liggitt commented:
> Sorry, not "add labels", but add something like SecretType which could distinguish if the configmap was created by controller automatically.

@deads2k any thoughts on the shape of this based on experience with the configmap-populating controller for the service serving cert controller?

WanLinghao commented:
/retest

WanLinghao commented:
/retest

WanLinghao commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kops-aws

WanLinghao commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kops-aws

WanLinghao commented:
/retest

WanLinghao commented:
@liggitt  @mikedanese  This patch has basically completed except the pending stuff.  Please take a glance.

WanLinghao commented:
@liggitt  @mikedanese  friendly ping

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -40,16 +40,16 @@ function run_kube_apiserver() {
-  # Enable features
-  ENABLE_FEATURE_GATES="DryRun=true"
+  # Features setting
+  FEATURE_GATES="DryRun=true,TokenRequest=false,TokenRequestProjection=false"
```
why was this change required?

WanLinghao replied:
Without this change, error would happen in 
https://github.com/kubernetes/kubernetes/blob/8012b9583e5301c2f66db4118b38433d035b87c4/test/cmd/get.sh#L134
https://github.com/kubernetes/kubernetes/blob/8012b9583e5301c2f66db4118b38433d035b87c4/test/cmd/core.sh#L28
Since test-cmd.sh is mainly test the function of kubectl, I think just disable TokenRequest feature to avoid this is reasonable.

liggitt replied:
if the apiserver isn't started with a private key for generating tokens (which this one is not), it should not serve the token request subresource, and the serviceaccount admission plugin should not inject tokenprojection volumes. we should not need to flip the feature gates differently here, but should instead make sure these features will work properly as the feature gates are promoted. the test you reference ("no configmaps exist in the namespace") is a fragile test that should probably be changed to more specifically test the precondition we actually care about.

liggitt commented on pkg/controller/serviceaccount/ca_configmap_controller.go:
```diff
@@ -0,0 +1,203 @@
+	"k8s.io/kubernetes/pkg/util/metrics"
+)
+
+const CAConfigMapName = "serviceaccount-ca-cert"
```
@mikedanese were you anticipating pinning a configmap name for this, or letting consumers indicate they wanted a configmap populated with the CA via a label or annotation?

cc @deads2k for his thoughts given similar experience with a ca-populating controller

mikedanese replied:
I strongly prefer pinning a name. However, I'd like it to indicate that it is the cluster's root certificate. It's somewhat orthogonal to service accounts. What about "kube-root-ca.crt"?

liggitt commented on pkg/controller/serviceaccount/ca_configmap_controller.go:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	cmInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: e.configMapDeleted,
```
what about an update that removes the CA key?

liggitt commented on pkg/controller/serviceaccount/ca_configmap_controller.go:
```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		return err
+	}
+	if ns.Status.Phase != v1.NamespaceActive {
```
most controllers don't check the namespace phase... they just try to work with the resources inside the namespace. is there a reason we need to for this controller?

liggitt commented on pkg/controller/serviceaccount/ca_configmap_controller.go:
```diff
@@ -0,0 +1,203 @@
+
+func (c *CaConfigMapController) namespaceUpdated(oldObj interface{}, newObj interface{}) {
+	newNamespace := newObj.(*v1.Namespace)
+	c.queue.Add(newNamespace.Name)
```
if you wanted to check the namespace phase and avoid enqueuing, you could do so here

liggitt commented on pkg/controller/serviceaccount/ca_configmap_controller.go:
```diff
@@ -0,0 +1,203 @@
+		return err
+	}
+
+	if _, err := c.client.CoreV1().ConfigMaps(ns.Name).Create(&c.configMap); err != nil && !apierrs.IsAlreadyExists(err) {
```
make a copy of the template configmap before passing to Create()... serialization mutates objects and we don't want races with other worker threads

liggitt commented on plugin/pkg/admission/serviceaccount/admission.go:
```diff
@@ -180,6 +182,7 @@ func (s *serviceAccount) Admit(a admission.Attributes) (err error) {
 		}
 	}
 
+	s.mountVolumeForTokenRequest(pod, a)
```
needs to be gated based on the feature-gate and the apiserver surfacing the token request API endpoint

liggitt replied:
AutomountServiceAccountToken should apply whether we are mounting a secret or a projection volume. I would not expect to inject both a secret and projection volume, only one or the other

liggitt commented on plugin/pkg/admission/serviceaccount/admission.go:
```diff
@@ -282,6 +285,51 @@ func (s *serviceAccount) enforceMountableSecrets(serviceAccount *corev1.ServiceA
+	for _, vol := range pod.Spec.Volumes {
+		if vol.Projected != nil {
+			for _, pro := range vol.Projected.Sources {
+				if pro.ServiceAccountToken != nil {
```
this seems to be locating and appending ca and namespace to an existing service account token projected volume, which is not what I expected. there are use cases for injecting a token without the namespace and CA.

I would only expect us to inject namespace and CA for the volume we are setting up to replace the existing service account token (the one we set up inside `mountServiceAccountToken`)

WanLinghao replied:
@liggitt  We can't mount two volumes to the same mount path. We need to make a choice here:

1. Create a independent volume which contains namespace&ca.crt projection volume, then mount it to the sub path of the path which token was mounted. 

2.  Create a independent volume which contains namespace and ca,crt projection volume, then 
mount it to a fix path like "/var/run/secrets/kubernetes.io/". 

3. Append ca.crt and namespace projection volume to the existing service account token projected volume just as my patch did.

I think for option 1&2, it will break current situation that ca.crt, namespace, token file appear in same path. And it is very hard to do so if user want. So maybe option3 is better choice. For the concern you proposed, if user  doesn't need namespace and CA, we just give up appending ca&namespace projection volume.
WDYT

liggitt replied:
> Append ca.crt and namespace projection volume to the existing service account token projected volume just as my patch did

This admission plugin is responsible for creating the volume to be mounted at `/var/run/secrets/kubernetes.io/serviceaccount`. It should either create a volume referencing a service account token secret (like it does today), or (when the projection feature is on and it detects the apiserver is serving tokenrequest endpoints) it should create a projection volume containing the service account token, ca.crt and namespace projections. It should not modify pre-existing projected volumes it finds.

As an optimization, if there is a pre-existing projection volume that perfectly matches the shape of the volume it would add, it can reuse it instead of appending a new volume (just like the code in master does today if there is an existing secret volume referencing the service account token secret).

mikedanese commented on pkg/controller/serviceaccount/ca_configmap_controller.go:
```diff
@@ -0,0 +1,203 @@
+limitations under the License.
+*/
+
+package serviceaccount
```
I think I would prefer if this was in a new directory in pkg/controller/certificates

WanLinghao replied:
directory pkg/controller/certificates already exists, what about pkg/controller/sa-certificates or maybe pkg/controller/serviceaccount/certificates?

mikedanese commented:
I prototyped something similar here:

https://github.com/mikedanese/kubernetes/tree/svcacct-admission

I wonder if we can do this incrementally by migrating the service account volumes to a projected volume now, but still using the old secret for the token. Regardless, can we separate out the admission changes and implement the controller in this PR?

WanLinghao commented:
@mikedanese  OK , I am working on it

WanLinghao committed:
To inject ca.crt into container when projected volume was specified, configmap should be created in each namespace.
This patch add a controller called "root-ca-cert-publisher" to complete above job as well as some bootstrap rbac policies.

WanLinghao commented:
@mikedanese @liggitt  Now this patch focus on configmap controller only. And all comments about it has been addressed.PTAL. 
 I will start another patch to solve admission controller change.

WanLinghao commented:
@mikedanese  friendly ping

WanLinghao commented on pkg/controller/crtconfigmap/crt_configmap_controller.go:
```diff
@@ -0,0 +1,261 @@
+		return
+	}
+	if newConfigMap.Name != CrtConfigMapName {
+		return
```
I don't think we should log a warning here since this could happen when user want to update any configmap besides ca configmap. 
Did you think of it as:
`if newConfigMap.Name != oldConfigMap.Name`

WanLinghao commented:
@mikedanese  comments addressed, PTAL

mikedanese commented on pkg/controller/certificates/rootcacertpublisher/root_ca_cert_publisher.go:
```diff
@@ -0,0 +1,230 @@
+		return
+	}
+
+	if !reflect.DeepEqual(c.configMap.Data, newConfigMap.Data) {
```
invert this check, so 
```go
if reflect.DeepEqual(c.configMap.Data, newConfigMap.Data) {
     return
```

mikedanese commented on test/cmd/apps.sh:
```diff
@@ -369,7 +369,8 @@ run_deployment_tests() {
   kube::test::get_object_assert deployment "{{range.items}}{{$id_field}}:{{end}}" 'nginx-deployment:'
-  kube::test::get_object_assert configmap "{{range.items}}{{$id_field}}:{{end}}" 'test-set-env-config:'
+  #configmap is special here due to controller will create kube-root-ca.crt for each namespace automatically
+  kube::test::get_object_assert 'configmaps/test-set-env-config' "{{$id_field}}" 'test-set-env-config'
```
I don't think the test here is necessary. These tests are used primarily to verify kubectl.

mikedanese commented on test/cmd/core.sh:
```diff
@@ -25,7 +25,7 @@ run_configmap_tests() {
   kube::log::status "Testing configmaps"
   kubectl create -f test/fixtures/doc-yaml/user-guide/configmap/configmap.yaml
-  kube::test::get_object_assert configmap "{{range.items}}{{$id_field}}{{end}}" 'test-configmap'
+  kube::test::get_object_assert 'configmap/test-configmap' "{{$id_field}}" 'test-configmap'
```
Same as above. I don't think the test here is necessary. These tests are used primarily to verify kubectl.

mikedanese commented on test/e2e/scheduling/resource_quota.go:
```diff
@@ -365,6 +365,22 @@ var _ = SIGDescribe("ResourceQuota", func() {
 	})
 
 	It("should create a ResourceQuota and capture the life of a configMap.", func() {
+		found, unchanged := 0, 0
```
Did this test begin to fail? As above, I don't think this is the right suite to test the ca.crt configmap publisher. It doesn't roll up to the right owners if this suite starts to fail. Something in sig-auth would be better.

mikedanese commented:
Can you make this disabled by default? I would like to tie enabling this controller to enabling the new service account token volume feature. See here:

https://github.com/kubernetes/kubernetes/blob/cbb422c214dd70735258b8e0af3b721613c2f188/cmd/kube-controller-manager/app/controllermanager.go#L333-L336

mikedanese commented:
@awly FYI

This almost looks good to me :) few more nits.

/approve

WanLinghao commented:
/test pull-kubernetes-e2e-kops-aws

WanLinghao commented:
@mikedanese The modify in test file is not to test the controller we added, but to prevent error happens. Since the test I modify all assumes that no configmap exists in a newly-create namespace. adding ca.crt configmap controller breaks this obviously.

WanLinghao commented:
@mikedanese all comments addressed , to make it easier to review, I didn't squash those two commits. I will squash them later if you feels good about this patch

awly commented on pkg/controller/certificates/rootcacertpublisher/root_ca_cert_publisher.go:
```diff
@@ -0,0 +1,232 @@
+				Name: RootCACertCofigMapName,
+			},
+			Data: map[string]string{
+				"ca.crt": string(rootCA),
```
It bothers me that this is immutable.
It would be awesome to support cluster CA rotation at some point, maybe this can be a callback?

WanLinghao replied:
We could just modify it when we want to support CA rotation in the future. As first version, I think it's better to keep the code simple and easy to understand. WDYT

mikedanese replied:
Rotation with a controller-manager restart is supported with this ¯\_(ツ)_/¯

awly commented on pkg/controller/certificates/rootcacertpublisher/root_ca_cert_publisher.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+// RootCACertPublisher manages certificate ConfigMap objects inside Namespaces
+type RootCACertPublisher struct {
```
https://github.com/golang/go/wiki/CodeReviewComments#package-names

WanLinghao replied:
Done

WanLinghao commented:
/test pull-kubernetes-integration

mikedanese commented on cmd/kube-controller-manager/app/certificates.go:
```diff
@@ -120,3 +123,33 @@ func startCSRCleanerController(ctx ControllerContext) (http.Handler, bool, error
+	sac, err := rootcacertpublisher.NewPublisher(
+		ctx.InformerFactory.Core().V1().ConfigMaps(),
+		ctx.InformerFactory.Core().V1().Namespaces(),
+		ctx.ClientBuilder.ClientOrDie("root-ca-crt-publisher"),
```
root-ca-cert-publisher, missing an e

mikedanese commented:
> The modify in test file is not to test the controller we added, but to prevent error happens. Since the test I modify all assumes that no configmap exists in a newly-create namespace. adding ca.crt configmap controller breaks this obviously.

Thanks for clarifying! SGTM.

mikedanese commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:
```diff
@@ -1031,6 +1031,31 @@ items:
+    creationTimestamp: null
+    labels:
+      kubernetes.io/bootstrapping: rbac-defaults
+    name: system:controller:root-ca-crt-publisher
```
@kubernetes/sig-auth-pr-reviews FYI on new controller permissions.

mikedanese commented:
/lgtm
/hold cancel

WanLinghao commented:
/test pull-kubernetes-e2e-kops-aws

mikedanese commented:
/retest

tedyu commented on pkg/controller/certificates/rootcacertpublisher/root_ca_cert_publisher.go:
```diff
@@ -0,0 +1,232 @@
+)
+
+// RootCACertCofigMapName is name of the configmap which stores certificates to access api-server
+const RootCACertCofigMapName = "kube-root-ca.crt"
```
typo: RootCACertCofigMapName -> RootCACertConfigMapName

tedyu commented on pkg/controller/certificates/rootcacertpublisher/root_ca_cert_publisher.go:
```diff
@@ -0,0 +1,232 @@
+
+	switch _, err := c.cmLister.ConfigMaps(ns.Name).Get(c.configMap.Name); {
+	case err == nil:
+		return nil
```
Why would we return for this case ?

WanLinghao replied:
Since if no error happens here, we already got a configmap, so no need to create one, just return

tedyu replied:
For the error case, a few lines below, we also return.
What's the purpose of the code following the switch ?

WanLinghao replied:
Here we need to assure each namespace got one configmap named "kube-root-ca.crt". The switch here means:

- No error indicates we already got "kube-root-ca.crt" configmap,nothing to do, return

- Not Found Error indicates we should go ahead and create one

- Other error, it means the GET operation itself got something wrong, it is not expected and we should return it
